core/shared/src/main/scala-3/cats/effect/cps.scala

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2021 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cats.effect
+
+import _root_.cps.{async, await, CpsAsyncMonad, CpsAwaitable, CpsMonad, CpsMonadPureMemoization}
+
+import cats.effect.kernel.{Async, Concurrent, Sync}
+
+import scala.util.Try
+
+object cps {
+
+  transparent inline def async[F[_]](using inline am: CpsMonad[F], F: Sync[F]): InferAsyncArg[F] =
+    new InferAsyncArg[F]
+
+  final class InferAsyncArg[F[_]](using am: CpsMonad[F], F: Sync[F]) {
+    transparent inline def apply[A](inline expr: A) =
+      F.defer(_root_.cps.Async.transform[F, A](expr)(using am))
+  }
+
+  final implicit class AwaitSyntax[F[_], A](val self: F[A]) extends AnyVal {
+    transparent inline def await(using inline am: CpsAwaitable[F]): A =
+      _root_.cps.await[F, A](self)
+  }
+
+  implicit def catsEffectCpsMonadPureMemoization[F[_]](implicit F: Concurrent[F]): CpsMonadPureMemoization[F] =
```

Comment:
As far as I understand, this construct is used conjointly with a `given` import to enable automatic colouring. I've already expressed my concerns against it, and I'd rather we avoided becoming enablers of (what I consider) a bad pattern. 

So I'd like this instance to be removed. If people really want it, they can add it in userland.

Reply:
No, automatic coloring is enabled only if given `cps.automaticColoring.tag`  is in scope.   I.e. it is always enabled on userland unless you do not reexport automatic coloring (not memorization) in your code.  So, define memorization here is safe.

Reply:
> it is always enabled on userland unless you do not reexport automatic coloring (not memorization) in your code. 

you mean it is always disabled in userland, right (unless the import is in scope which requires decision from the user) ?

Even if that's the case, I still oppose the auto-colouring pattern and wouldn't want this library (cats-effect-cps) to make it easier for people to enable.

Reply:
yes.  
  Hmm, a problem that memoization is a right description on functionality and disabling it potentially can break other functionality, not related to coloring at all.

  The right way to disable automatic coloring will be to define another tag to force implicit resolution conflict.  We can make AutomaticColoringTag  be depended on monad,  so you can reexport something like:
```
@implicitAmbiguous ("Automatic coloring for IO is disabled, use another library if you want automatic coloring")
given DisableAutomaticColoringForIO:  cps.automaticColoring.AutomaticColoringTag[IO]
```

Reply:
> Hmm, a problem that memoization is a right description on functionality and disabling it potentially can break other functionality, not related to coloring at all.

I see, thanks 😄 . out of curiosity, could you formulate an example of how memoization is used in contexts other than auto-colouring ? In particular, I want to ensure that its presence wouldn't break least-surprise.

Reply:
Right now, it's not used in dotty-cps-async for purposes other than coloring.  (It's why I have written `potentially`).  But imagine you, as a developer,  building something which should logically works with memoization. For example, it can be a cache layer in some service or broadcasting events to listeners, where events are processed once, but the results are broadcasted.   And then you are discovering that this is impossible to do if you import cps-async structures for IO via cats-effect-cps, because authors of cats-effect-cps dislike automatic coloring ;).      
On the other side, it's your choice  - of course, in your library you can do whatever you want ;)

// btw, I'm not happy with the current automatic coloring for effect monads and trying to explore now some other ways, which should be more concise for developers.  Some of those approaches use memoization more explicitly than now.   I will be able to say more a bit later when will have some results.

Reply:
@Baccata I believe the memoization is needed for correctly translating `val`s under some circumstances. This is one of the areas where monadless can sometimes change the meaning of the code. The trick is that dotty-cps-async isn't *running* the effects, but rather performing a transformation of the call graph. Thus, in some scenarios, it needs the ability to repeatedly sequence the same effect multiple times but without re-evaluating it.

Reply:
> The trick is that dotty-cps-async isn't running the effects, but rather performing a transformation of the call graph. 

Yeah that is pretty clear, and it's quite amazing that it's able to do that on a reasonably large subset of the language (including while loops, etc) 

>  I believe the memoization is needed for correctly translating vals under some circumstances

I do understand (intuitively) that memoization is useful when auto-colouring is involved, but the author has indicated (in the message above yours) that memoization is, at the moment, only used in that very context. Do you have an example in mind of a possible transformation (even one not currently supported) that would benefit from memoization when the user has to manually apply colouring ? (I'm not being rhetorical here, genuinely curious as to whether you can think of one)

Reply:
> I do understand (intuitively) that memoization is useful when auto-colouring is involved, but the author has indicated (in the message above yours) that memoization is, at the moment, only used in that very context. Do you have an example in mind of a possible transformation (even one not currently supported) that would benefit from memoization when the user has to manually apply colouring ? (I'm not being rhetorical here, genuinely curious as to whether you can think of one)
 
Ah, it would help if I could read. You're correct.

So I defined the instance because it fails to compile *without* it. That's the baseline, but beyond that, the scenarios I'm thinking about would have to involve something like `val x = ioa.await` followed by some weird construct which defies reconstitution into a `flatMap`. I'm really not sure what that would be though, if anything.

Either way, I'm not *too* worried about it. I agree we don't want autocoloring, but since that's controlled independently, implementing this abstraction seems harmless enough. `Concurrent[F].memoize` is safe to use.

Reply:
> So I defined the instance because it fails to compile without it

That is a sufficient argument for me to accept the presence of the instance in this PR 👍